src/main/java/org/opensearch/security/resources/ResourceAccessHandler.java

```diff
@@ -0,0 +1,560 @@
+
+        LOGGER.debug("Deleting all resource sharing records for user {}", user.getName());
+
+        resourceSharingIndexHandler.deleteAllRecordsForUser(user.getName(), ActionListener.wrap(listener::onResponse, exception -> {
+            LOGGER.error(
+                "Failed to delete all resource sharing records for user {}: {}",
+                user.getName(),
+                exception.getMessage(),
+                exception
+            );
+            listener.onFailure(exception);
+        }));
+    }
+
+    /**
+     * Loads all resources within the specified resource index.
+     *
+     * @param resourceIndex The resource index to load resources from.
+     * @param listener      The listener to be notified with the set of resource IDs.
+     */
+    private void loadAllResources(String resourceIndex, ActionListener<Set<String>> listener) {
+        this.resourceSharingIndexHandler.fetchAllDocuments(resourceIndex, listener);
+    }
+
+    /**
+     * Loads resources owned by the specified user within the given resource index.
+     *
+     * @param resourceIndex The resource index to load resources from.
+     * @param userName      The username of the owner.
+     * @param listener      The listener to be notified with the set of resource IDs.
+     */
+    private void loadOwnResources(String resourceIndex, String userName, ActionListener<Set<String>> listener) {
+        this.resourceSharingIndexHandler.fetchDocumentsByField(resourceIndex, "created_by.user", userName, listener);
+    }
+
+    /**
+     * Loads resources shared with the specified entities within the given resource index, including public resources.
+     *
+     * @param resourceIndex The resource index to load resources from.
+     * @param entities      The set of entities to check for shared resources.
+     * @param recipient     The type of entity (e.g., users, roles, backend_roles).
+     * @param listener      The listener to be notified with the set of resource IDs.
+     */
+    private void loadSharedWithResources(
+        String resourceIndex,
+        Set<String> entities,
+        String recipient,
+        ActionListener<Set<String>> listener
+    ) {
+        Set<String> entitiesCopy = new HashSet<>(entities);
+        // To allow "public" resources to be matched for any user, role, backend_role
+        entitiesCopy.add("*");
+        this.resourceSharingIndexHandler.fetchDocumentsForAllActionGroups(resourceIndex, entitiesCopy, recipient, listener);
+    }
+
+    /**
+     * Checks if the given resource is owned by the specified user.
+     *
+     * @param document The ResourceSharing document to check.
+     * @param userName The username to check ownership against.
+     * @return True if the resource is owned by the user, false otherwise.
+     */
+    private boolean isOwnerOfResource(ResourceSharing document, String userName) {
+        return document.getCreatedBy() != null && document.getCreatedBy().getCreator().equals(userName);
```

Comment:
Question/concern here is whether username would be guaranteed to be unique, specifically when considering multiple SSO providers/internal users?